torch/_refs/__init__.py

```diff
@@ -1658,49 +1658,135 @@ def item(a: TensorLikeType) -> NumberType:
     return number_type(prims.item(a))
 
 
-# TODO: extend to device and layout similar to
-# aten/src/ATen/native/TensorConversions.cpp:to_will_alias
+# fast path when `to` returns an alias to input. This mimics the same function in aten
 def _to_will_alias(
     a: TensorLikeType,
-    dtype: torch.dtype,
-    copy: bool,
-    memory_format: torch.memory_format,
+    device: Optional[torch.device],
+    dtype: Optional[torch.dtype],
+    copy: Optional[bool],
+    memory_format: Optional[torch.memory_format],
+    layout: Optional[torch.layout],
 ) -> bool:
     return (
         not copy
-        and a.dtype == dtype
+        and (device is None or a.device == device)
+        and (dtype is None or a.dtype == dtype)
+        and (layout is None or a.layout == layout)
         and (
-            memory_format == torch.preserve_format
-            or a.is_contiguous(memory_format=memory_format)
+            memory_format is None
+            or memory_format == torch.preserve_format
+            or utils.is_contiguous_for_memory_format(a, memory_format=memory_format)
         )
     )
 
 
-# TODO: there are multiple overloads of to, but we only support one here
-# aten/src/ATen/native/native_functions.yaml lists all overloads
-# Missing overloads:
-# to.dtype_layout
-# to.device
-# to.other
-# https://github.com/pytorch/pytorch/issues/84264
-def to(
+def _to_dispatch(
     a: TensorLikeType,
-    dtype: torch.dtype,
-    non_blocking: bool = False,
-    copy: bool = False,
-    memory_format: torch.memory_format = torch.preserve_format,
-) -> TensorLikeType:
-    if _to_will_alias(a, dtype, copy, memory_format):
+    *args,
+    device: Optional[torch.device] = None,
+    dtype: Optional[torch.dtype] = None,
+    non_blocking: Optional[bool] = None,
+    copy: Optional[bool] = None,
+    memory_format: Optional[torch.memory_format] = None,
+    layout: Optional[torch.layout] = None,
+    pin_memory: Optional[bool] = None,
+):
+    arg_list = list(args)
+    arg0, *arg_list = (
+        arg_list
+        if arg_list
+        else [
+            None,
+        ]
+    )
+    arg0 = torch.device(arg0) if isinstance(arg0, str) else arg0
+    if isinstance(arg0, TensorLike):
+        # overload to `to.other`
+        device = arg0.device
+        dtype = arg0.dtype
+        layout = arg0.layout
+        # TODO: is_pinned is not currently supported in refs or fake_tensor
+        # https://github.com/pytorch/pytorch/issues/84925
+        # pin_memory = arg0.is_pinned()
+        # load positional arg_listuments:
+        if arg_list:
+            assert non_blocking is None
```

Comment:
This pattern seems fine. One easier alternative might be to define helper functions like `_to_tensor` with the correct signature and pass `*args` and `**kwargs` to them and have them return a dict of the correct metadata

Reply:
(If anyone ever has to refactor this code, that's the approach I imagine they'll take)

Reply:
(It might be interesting to look at how `bind` is used here:

https://github.com/pytorch/pytorch/blob/706b99030656c573619cebaa3be9298a575fc776/torch/_prims_common/wrappers.py#L99

)

Reply:
! Sorry about my sloppy python~
I was sure how to handle kwargs binding cleanly in my earlier attempt so ended up with them explicit in the signature. I'll make another attempt to switch to inspect.signature.

Reply:
I need to think a little bit more about this. If we can magically have all arguments converted to a proper dictionary, then we would not have to worry about doing this manual~ish dispatch.

But the tricky part is that each of `torch.ops.aten.to.XXX` overload takes arguments slightly differently and they are implicit in the sense that they all takes `*args, **kwargs` and `inspect` can't really do anything useful there.

> One easier alternative might be to define helper functions like _to_tensor with the correct signature

^^^ I don't know how we can have one `_to_tensor` signature and resolve the `*args`. To figure out the potential `*args` dispatch, I think we'll need three `_to_tensor_XXX` variants and use that with `inspect` to automatically sort out the signature. But that seems to bring us the full circle back to the original implementation, where we are just if/elif on the input arguments. :laughing: 

Maybe I'm missing something obvious, I can use some more elaborated comments. :bow:

Reply:
Here's one suggestion on how to handle dispatch. At least one positional argument must be passed to `_to_dispatch` in this example but everything else can be passed as either a positional or a keyword argument.

```py
import torch
from functools import singledispatch

@singledispatch
def _to_dispatch(*args, **kwargs):
    raise NotImplementedError

@_to_dispatch.register
def _to_device(device: torch.device, dtype, non_blocking, copy, memory_format):
    print("to.device overload!")
    layout = None
    pin_memory = None
    return device, dtype, non_blocking, copy, memory_format, layout, pin_memory

@_to_dispatch.register
def _to_device(device: str, dtype, non_blocking, copy, memory_format):
    print("to.device str overload!")
    device = torch.device(device)
    return _to_dispatch(device, dtype, non_blocking, copy, memory_format)

@_to_dispatch.register
def _to_dtype(dtype: torch.dtype, non_blocking, copy, memory_format):
    print("to.dtype overload!")
    layout = None
    pin_memory = None
    device = None
    return device, dtype, non_blocking, copy, memory_format, layout, pin_memory

@_to_dispatch.register
def _to_other(other: torch.Tensor, non_blocking, copy, memory_format):
    print("to.other overload!")
    layout = other.layout
    pin_memory = other.is_pinned()
    device = other.device
    dtype = other.dtype
    return device, dtype, non_blocking, copy, memory_format, layout, pin_memory
```

Reply:
Thanks a lot for hand-holding me through this :smile: 

I went with singledispatch that returns a dictionary for kwargs. I feel this is a relatively straightforward way to handle this messy overload. Feels like this also works more naturally with the suggestion @mruberry has on canonicalize the kwargs per attributes from input tensor.